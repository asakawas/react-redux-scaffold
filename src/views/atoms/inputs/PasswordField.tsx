import React, { FunctionComponent } from 'react';
import MuiTextField, { TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';

/**
 * The type of props of TextField.
 */
export type TextFieldProps = Readonly<
  Required<Pick<MuiTextFieldProps, 'name' | 'label'>> &
    Pick<MuiTextFieldProps, 'defaultValue' | 'fullWidth' | 'inputRef'> & {
      /** If true or omitted, this input become read-only. */
      readOnly?: boolean;
    }
>;

const TextField: FunctionComponent<TextFieldProps> = ({
  name,
  label,
  defaultValue,
  fullWidth,
  inputRef,
  readOnly,
}) => (
  <MuiTextField
    name={name}
    id={name}
    label={label}
    defaultValue={defaultValue}
    InputProps={{
      readOnly,
      inputProps: {
        'aria-label': `input-${name}`,
      },
    }}
    type="password"
    autoComplete="current-password"
    fullWidth={fullWidth}
    inputRef={inputRef}
    helperText={readOnly ? 'Read-only' : undefined}
  />
);

export default React.memo(TextField);
