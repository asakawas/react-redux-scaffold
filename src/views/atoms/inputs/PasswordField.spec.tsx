import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import PasswordField from './PasswordField';

describe('PasswordField', () => {
  test('is rendered as expected', () => {
    const { asFragment } = render(<PasswordField name="test" label="Test" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('readOnly ver is rendered as expected', () => {
    const { asFragment } = render(<PasswordField name="test" label="Test" readOnly />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('has password attributes', () => {
    const renderResult = render(<PasswordField name="test" label="Test" />);
    const tree = renderResult.getByLabelText('input-test');

    expect(tree).toHaveAttribute('type', 'password');
    expect(tree).toHaveAttribute('autocomplete', 'current-password');
  });

  test('can be typed in if readOnly is false', async () => {
    const renderResult = render(<PasswordField name="test" label="Test" />);

    const tree = renderResult.getByLabelText('input-test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
    await userEvent.type(tree, 'test test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('test test');
  });

  test('can not be typed in if readOnly is true', async () => {
    const renderResult = render(<PasswordField name="test" label="Test" readOnly />);

    const tree = renderResult.getByLabelText('input-test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
    await userEvent.type(tree, 'test test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
  });

  test("doesn't has a helper text to show 'Read-only' if readOnly is false", () => {
    const renderResult = render(<PasswordField name="test" label="Test" />);

    const tree = renderResult.queryByText('Read-only');
    expect(tree).toBeNull();
  });

  test('has a helper text to show "Read-only" if readOnly is true', () => {
    const renderResult = render(<PasswordField name="test" label="Test" readOnly />);

    const tree = renderResult.queryByText('Read-only');
    expect(tree).not.toBeNull();
  });
});
