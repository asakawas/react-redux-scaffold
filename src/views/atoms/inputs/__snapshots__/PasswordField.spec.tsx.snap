// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField is rendered as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="test"
      id="test-label"
    >
      Test
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        aria-label="input-test"
        autocomplete="current-password"
        class="MuiInputBase-input MuiInput-input"
        id="test"
        name="test"
        type="password"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PasswordField readOnly ver is rendered as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="test"
      id="test-label"
    >
      Test
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-describedby="test-helper-text"
        aria-invalid="false"
        aria-label="input-test"
        autocomplete="current-password"
        class="MuiInputBase-input MuiInput-input"
        id="test"
        name="test"
        readonly=""
        type="password"
        value=""
      />
    </div>
    <p
      class="MuiFormHelperText-root"
      id="test-helper-text"
    >
      Read-only
    </p>
  </div>
</DocumentFragment>
`;
